node/src/components/consensus/era_supervisor.rs

```diff
@@ -1269,6 +1272,39 @@ where
     }
 }
 
+async fn execute_finalized_block<REv>(
```

Comment:
I agree that the fact this code has to get the deploys is a smell.

Here is my proposal (if you agree, we can make a ticket):

We change `EffectBuilder::validate_block` to be:

```rust
async fn validate_block<I>(self, sender: I, block: ProposedBlock) -> Result<ValidatedBlock, BlockValidationError>
```

Where:

```rust
struct ValidatedBlock {
   proposed_block: ProposedBlock,
   deploys: Deploys
}
```

Replay detection logic would move into the block validator and be rolled up into `BlockValidationError`.

This would improve type correctness.  Any method that should operate on a `ValidatedBlock`, rather than a merely `ProposedBlock`, will have to say-so in its type signature.

While it's true that `deploys` would still be `ValidatedBlock`, it's an implementation detail.  If we want, we could put `ValidatedBlock` in `linear_chain.rs` and make `ValidatedBlock::deploys` a private method.

WDYT?

Reply:
I like the idea of a `ValidatedBlock` type — similar to what we do with vertices in Highway. It would also require a change to the Highway code, however: Currently the `ProposedBlock` is part of the Highway unit wire format, but in the protocol state we'd want to use `ValidatedBlock`. Anyway, that's probably something for a different PR.

I'd still like to move everything that uses `Deploy` or `DeployHash` out of `consensus`, however.

Reply:
Okay, this will be moved out of consensus and into `LinearChain` in an upcoming PR I am working on.